src/com/ms/silverking/cloud/gridconfig/GridConfiguration.java

```diff
@@ -15,164 +15,165 @@
+    } else {
+      defaultBase = null;
     }
-    
-    static void staticInit() {
-        String    defaultBaseVal;
-        String    envSuffixVal;
-        
-        defaultBaseVal = getFromEnvOrProperty(defaultBaseEnvVar, defaultBaseProperty);
-        if (defaultBaseVal != null && defaultBaseVal.trim().length() > 0) {
-            defaultBase = new File(defaultBaseVal);
-        } else {
-            defaultBase = null;
-        }
-        
-        defaultGC = getFromEnvOrProperty(defaultGCEnvVar, defaultGCProperty);
-        
-        envSuffixVal = PropertiesHelper.systemHelper.getString(envSuffixProperty, PropertiesHelper.UndefinedAction.ZeroOnUndefined);
-        if (envSuffixVal == null) {
-            envSuffixVal = PropertiesHelper.envHelper.getString(envSuffixEnvVar, defaultEnvSuffix);
-        }
-        envSuffix = envSuffixVal;
+
+    defaultGC = getFromEnvOrProperty(defaultGCEnvVar, defaultGCProperty);
+
+    envSuffixVal = PropertiesHelper.systemHelper.getString(envSuffixProperty,
+        PropertiesHelper.UndefinedAction.ZeroOnUndefined);
+    if (envSuffixVal == null) {
+      envSuffixVal = PropertiesHelper.envHelper.getString(envSuffixEnvVar, defaultEnvSuffix);
     }
-    
-    public static File getDefaultBase() {
-        if (defaultBase == null) {
-            Log.warning("defaultBase is undefined");
-            Log.warningf("Set env var %s or property %s", defaultBaseEnvVar, defaultBaseProperty);
-            throw new RuntimeException("defaultBase is undefined");
+    envSuffix = envSuffixVal;
+  }
+
+  public static File getDefaultBase() {
+    if (defaultBase == null) {
+      Log.warning("defaultBase is undefined");
+      Log.warningf("Set env var %s or property %s", defaultBaseEnvVar, defaultBaseProperty);
+      throw new RuntimeException("defaultBase is undefined");
+    } else {
+      return defaultBase;
+    }
+  }
+
+  public static String getDefaultGC() {
+    return defaultGC;
+  }
+
+  public GridConfiguration(String name, Map<String, String> envMap) {
+    this.name = name;
+    this.envMap = envMap;
+  }
+
+  public static GridConfiguration parseFile(File gcBase, String gcName) throws IOException {
+    Map<String, String> envMap;
+
+    if (gcName == null) {
+      if (defaultGC != null) {
+        gcName = defaultGC;
+      } else {
+        throw new RuntimeException("No GC defined. Missing -g ?");
+      }
+    }
+    envMap = readEnvFile(new File(gcBase, gcName + envSuffix));
+    return new GridConfiguration(gcName, envMap);
+  }
+
+  public static GridConfiguration parseFile(String gcName) throws IOException {
+    return parseFile(getDefaultBase(), gcName);
+  }
+
+  public static Map<String, String> readEnvFile(File envFile) throws IOException {
+    Map<String, String> envMap;
+    String line;
+    BufferedReader reader;
+
+    envMap = new HashMap<String, String>();
+    reader = new BufferedReader(new InputStreamReader(new FileInputStream(envFile)));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *reader*. The resource is referenced in statements at the following lines: 119, 121. The resource closure statement is at line: 134. There are other execution paths that do not contain closure statements, for example, when BufferedReader.readLine throws an exception. Either a) close *reader* in a try-finally block or b) close the resource by declaring *reader* in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).